gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaaas_driver.py

```diff
@@ -0,0 +1,418 @@
+
+
+class VPNAgentNotFound(exceptions.NeutronException):
+    message = _("VPN Agent not found in agent_db")
+
+
+class IpsecSiteConnOverlappingPeerAdress(exceptions.NeutronException):
+    message = _("Overlapping peer cidr '%(peer_cidr)s' for \
+        peer '%(peer_address)s'")
+
+
+class DuplicateVpnService(exceptions.NeutronException):
+    message = _("Duplicate vpn service '%(name)s'")
+
+
+class VpnServiceInvalidSubnet(exceptions.NeutronException):
+    message = _("Invalid subnet '%(subnet_id)s'")
+
+
+class InvalidVPNServiceDesc(exceptions.NeutronException):
+    message = _("Invalid vpn svc desc '%(desc)s' should be \
+        e.g., fip=192.168.20.120;tunnel_local_cidr=10.0.0.0/24")
+
+
+class VPNServiceInErrorState(exceptions.NeutronException):
+    message = _("VPN service is in error state")
+
+
+class VPNPluginExt(VPNPlugin, agentschedulers_db.AgentSchedulerDbMixin):
+    """
+    Extends the base VPN Plugin class to inherit agentdb too.
+    Required to get agent entry into the database.
+    """
+
+    def __init__(self):
+        super(VPNPluginExt, self).__init__()
+
+
+class OCIPsecVpnDriverCallBack(object):
+    """Callback for IPSecVpnDriver rpc."""
+
+    target = oslo_messaging.Target(version=BASE_IPSEC_VERSION)
+
+    def __init__(self, driver):
+        super(OCIPsecVpnDriverCallBack, self).__init__()
+        self.driver = driver
+
+    def create_rpc_dispatcher(self):
+        return n_rpc.PluginRpcDispatcher([self])
+
+    def update_status(self, context, **status):
+        """Update status of vpnservices."""
+        status = status['status']
+        if 'ipsec_site_connections' not in status[0]:
+            status[0]['ipsec_site_connections'] = {}
+        plugin = self.driver.service_plugin
+        plugin.update_status_by_agent(context, status)
+
+    def ipsec_site_conn_deleted(self, context, **resource_id):
+        """ Delete ipsec connection notification from driver."""
+        resource_id = resource_id['id']
+        plugin = self.driver.service_plugin
+        plugin._delete_ipsec_site_connection(context, resource_id)
+
+    def vpnservice_deleted(self, context, **kwargs):
+        vpnservice_id = kwargs['id']
+        plugin = self.driver.service_plugin
+        plugin._delete_vpnservice(context, vpnservice_id)
+
+    def update_ipsec_site_conn_description(self, context, conn_id,
+                                           description):
+        plugin = self.driver.service_plugin
+        plugin.update_ipsec_site_connection(context, conn_id, description)
+
+
+class OCIpsecVpnAgentApi(service_drivers.BaseIPsecVpnAgentApi):
+    """API and handler for OC IPSec plugin to agent RPC messaging."""
+    target = oslo_messaging.Target(version=BASE_IPSEC_VERSION)
+
+    def __init__(self, topic, default_version, driver):
+        super(OCIpsecVpnAgentApi, self).__init__(
+            topic, default_version, driver)
+
+    def _is_agent_hosting_vpnservice(self, agent, vpnservice_id):
+        """
```

Comment:
Why is vpnservice_id attribute unused in the function definition ?
The comment below is mentioning agents running on compute node !. This need to be corrected.